<launch>

  <arg name="isam2_node" default="isam2_vio"/>
  <arg name="images_compressed" default="true"/>
  
  <arg name="robot" default="minitaur"/>
  <arg name="calibration_file"
    default="$(find legged_vio)/config/zed_imu_camera_altered.yaml"/>
  <arg name="left_camera" default="/zed/zed_node/left/image_rect_color"/>
  <arg name="right_camera" default="/zed/zed_node/right/image_rect_color"/>
  
  <!-- Load calibration_file for any node outside of Image Processor  -->
  <rosparam command="load" file="$(arg calibration_file)"/>
  
  <group ns="$(arg robot)">
    
    <!-- Image Processor Nodelet  -->
    <!--  (NOTE: to print output, include output="screen" command)  -->
    <node pkg="nodelet" type="nodelet" name="image_processor"
      args="standalone legged_vio/ImageProcessorNodelet"> <!--output="screen"-->
      
      <rosparam command="load" file="$(arg calibration_file)"/>
      <param name="grid_row" value="4"/>
      <param name="grid_col" value="5"/>
      <param name="grid_min_feature_num" value="3"/>
      <param name="grid_max_feature_num" value="4"/>
      <param name="pyramid_levels" value="3"/>
      <param name="patch_size" value="15"/>
      <param name="fast_threshold" value="10"/>
      <param name="max_iteration" value="30"/>
      <param name="track_precision" value="0.01"/>
      <param name="ransac_threshold" value="3"/>
      <param name="stereo_threshold" value="5"/>

      <!-- This nodelet subscribes to imu, cam0_image, and cam1_image topics -->
      <!-- Remap to the topics that are published by zed camera: -->
      <remap from="~imu" to="/zed/zed_node/imu/data"/> 
      <remap from="~cam0_image" to="$(arg left_camera)"/>
      <remap from="~cam1_image" to="$(arg right_camera)"/>

    </node>
          
    <!-- Uncompress images if compressed  -->
    <node name="republish_left" type="republish" pkg="image_transport"
      args="compressed in:=$(arg left_camera) raw out:=$(arg left_camera)" 
      if ="$(eval arg('images_compressed'))"/>
    <node name="republish_right" type="republish" pkg="image_transport"
      args="compressed in:=$(arg right_camera) raw out:=$(arg right_camera)"
      if ="$(eval arg('images_compressed'))"/>
    
  </group>
  
  <!-- ISAM2 Node  -->
  <!--  (NOTE: to print output, include output="screen" command)  -->
  <node pkg="legged_vio" name="$(arg isam2_node)" type="$(arg isam2_node)" 
    required="true" output="screen"/>

</launch>

<?ignore
  <!-- START: unusued parameters intitially intended for isam2 -->
  <!-- These values should be standard deviation -->
  <param name="noise/gyro" value="0.005"/>
  <param name="noise/acc" value="0.05"/>
  <param name="noise/gyro_bias" value="0.001"/>
  <param name="noise/acc_bias" value="0.01"/>
  <param name="noise/feature" value="0.035"/>

  <param name="initial_state/velocity/x" value="0.0"/>
  <param name="initial_state/velocity/y" value="0.0"/>
  <param name="initial_state/velocity/z" value="0.0"/>

  <!-- These values should be covariance -->
  <param name="initial_covariance/velocity" value="0.25"/>
  <param name="initial_covariance/gyro_bias" value="0.01"/>
  <param name="initial_covariance/acc_bias" value="0.01"/>
  <param name="initial_covariance/extrinsic_rotation_cov" value="3.0462e-4"/>
  <param name="initial_covariance/extrinsic_translation_cov" value="2.5e-5"/>
  <!-- END: unusued parameters intitially intended for isam2 -->
?>
